tuner/tuner/candidate_gen.py

```diff
@@ -147,6 +147,50 @@ def get_default_output_dir() -> str:
     return "tuning_" + datetime.now().strftime("%Y_%m_%d_%H_%M")
 
 
+def determine_td_specs_to_link(
+    td_specs: list[ir.Module], warned_overlap_matchers: set[str]
+) -> tuple[list[ir.Module], set[str]]:
```

Comment:
Instead of having the caller provide the known duplicate matchers and returning new duplicated, I think it would be easier to add a boolean flag that decides whether to print a warning about duplicates. This simplifies the interface. If you want to test the duplicate finding logic, you can make it a helper function and test in isolation.